OFEP-single-context-paradigm.md

```diff
@@ -0,0 +1,105 @@
+## OFEP-single-context-paradigm
+
+## State: DRAFTING
+
+This draft outlines a "single-context" paradigm - an alternative pattern and set of APIs supporting client use-cases.
+
+## Background
+
+In contrast with server-side or other service-type applications, client side apps typically operate in the context of a single user.
+Most feature flagging SDKs for these applications have been designed with this in mind.
+In summary, most client/web SDKs operate something like this:
+
+- an initialization occurs, which fetches evaluated flags in bulk for a given context (user)
+- the evaluated flags are cached in the SDK
+- flag evaluations take place against this cache, without a need to provide context (context was already used to evaluate flags in bulk)
+- Functions/methods are exposed on the SDK that signal the cache is no longer valid, and must be reconciled based on a context change. This frequently involves a network request or I/O operation.
+
+This paradigm doesn't fit well with the existing SDK, which, like most server-side SDKs, emphasizes realtime evaluation and context-per-evaluation.
+
+Though not all client side SDKs function in this way, those that do allow context per evaluation can conform to this model fairly easily.
+
+## Design
+
+To better support the "single-context" paradigm we need to define some additional handlers and add some flexibility to current APIs.
+In short, the application author now sets context globally using the existing global context mutator on the OpenFeature API object.
+This single context is used by providers for initialization and fetching evaluated flag values in bulk.
+All evaluations functions are "context-less".
+When the context is changed, providers are signalled to update their cache of evaluated flags.
+
+### Provider changes
+
+#### On-context-set handler
+
+Providers may need a mechanism to understand when their cache of evaluated flags must be invalidated or updated. An `on-context-set` handler can be defined which performs whatever operations are needed to reconcile the evaluated flags with the new context. The OpenFeature SDK calls this handler when the global context is modified.
+
+```typescript
+
+interface Provider {
+  //...
+
+  // a handler called by the SDK when context is modified
+  onContextSet?(oldContext: EvaluationContext, newContext: EvaluationContext): Promise<void>
+
+  //...
+}
+```
+
+While the `on-context-set` handler is executing, the cache of resolved flags may be considered "stale". `Provider authors` and `application authors` should understand the consequences of evaluating flags in this state.
+
+#### Remove evaluation-context parameter on resolvers
+
+With all evaluation context represented in the global context, resolvers no longer need it as a parameter. Providers that require context at evaluation can simply access the global evaluation context for use in their evaluation.
```

Comment:
@fabriziodemaria yes, excellent point.

I think this question also brings up whether or not we want to maintain hooks the same way in the static context paradigm. _I think the answer is yes, but I'm open to other proposals._

In terms of your specific point, I think in languages where this is a concern, it would be necessary to do some locking to prevent unpredictable behavior. We have similar locks for similar programs in some server SDKs already.

Reply:
One proposal could be for the Provider's resolve methods to still accept the `ctx` from the SDK Client, rather than having the Provider internally read the global property; this way we can at least ensure that the same `ctx` is used in any one evaluation and its corresponding hooks. As a side note, I also think that keeping the `ctx` input argument in the Provider interface could streamline slightly writing tests for the Providers, at least in certain programming languages; on top of that, this would make the Provider's implementation perhaps slightly more agnostic of the SDK's global properties, which sounds a little cleaner to work with. What do you think?

Reply:
@fabriziodemaria I think that is a good idea because it _probably_ reduces the need for locking and gives us more control, and it also is more consistent, as you point out. I will make a change here and also on the draft [PR spec](https://github.com/open-feature/spec/pull/171).

Reply:
I've made this change here: https://github.com/open-feature/ofep/pull/41/commits/4083c83bd37ae4f37dc02155b27b256a6ac937bc